src/token.rs

```diff
@@ -662,80 +886,174 @@ impl Token {
-                }
+    fn clear_private_session_objects(&mut self) {
+        let mut priv_handles = Vec::<CK_OBJECT_HANDLE>::new();
+        for (handle, obj) in self.session_objects.iter() {
+            if obj.is_private() {
+                priv_handles.push(*handle);
             }
-            CKU_SO => {
-                if old.is_none() {
-                    return CKR_PIN_INCORRECT;
-                }
-                let ret =
-                    self.so_login.change_pin(&self.info, pin, old.unwrap());
-                if ret != CKR_OK {
-                    return ret;
-                }
-                /* update pin in storage */
-                match self.store_pin_object(
-                    SO_PIN_UID.to_string(),
-                    "SO PIN".to_string(),
-                    pin.clone(),
-                ) {
-                    Ok(()) => (),
-                    Err(_) => return CKR_GENERAL_ERROR,
-                }
+        }
+
+        /* remove all private session objects */
+        for handle in priv_handles {
+            let _ = self.session_objects.remove(&handle);
+            self.handles.remove(handle);
+        }
+    }
+
+    fn clear_session_objects(&mut self, session: CK_SESSION_HANDLE) {
+        let mut handles: Vec<CK_OBJECT_HANDLE> = Vec::new();
+        for (_, obj) in self.session_objects.iter() {
+            if obj.get_session() == session {
+                handles.push(obj.get_handle());
             }
-            _ => return CKR_GENERAL_ERROR,
         }
 
-        /* If we set a PIN it means we switched to require Logins */
-        self.info.flags |= CKF_LOGIN_REQUIRED;
+        for h in handles {
+            let _ = self.session_objects.remove(&h);
+            self.handles.remove(h);
+        }
+    }
 
-        self.update_token_pin_flags();
+    fn encrypt_value(&self, uid: &String, val: &Vec<u8>) -> KResult<Vec<u8>> {
+        if let Some(ref kek) = self.kek {
+            let aad = uid.as_bytes();
+            let tag_len = 8;
+            let mut iv = [0u8; 12];
+            get_random_data(&mut iv)?;
+            let mut params = CK_GCM_PARAMS {
+                pIv: iv.as_ptr() as *mut CK_BYTE,
+                ulIvLen: 12,
+                ulIvBits: 96,
+                pAAD: aad.as_ptr() as *mut CK_BYTE,
+                ulAADLen: aad.len() as CK_ULONG,
+                ulTagBits: (tag_len * 8) as CK_ULONG,
+            };
+            let mech: CK_MECHANISM = CK_MECHANISM {
+                mechanism: CKM_AES_GCM,
+                pParameter: &mut params as *mut interface::CK_GCM_PARAMS
+                    as *mut _,
+                ulParameterLen: std::mem::size_of::<CK_GCM_PARAMS>()
+                    as CK_ULONG,
+            };
+            let aes = self.mechanisms.get(CKM_AES_GCM)?;
+            let mut op = aes.encryption_new(&mech, &kek)?;
+            let mut clen: CK_ULONG =
+                op.encryption_len(val.len() as CK_ULONG)? as CK_ULONG + 12;
```

Comment:
another use of magic number 12 that should be replaced by `iv_len` or some shared value. There are two more on lines 953 and 957.

Reply:
I will add a const variable with a name that makes it clearer